pubtools/_quay/item_processor.py

```diff
@@ -0,0 +1,414 @@
+                raw=True,
+            )
+            digests.append(
+                ManifestArchDigest(
+                    manifest=manifest,
+                    digest=arch_manifest["digest"],
+                    arch=arch_manifest["platform"]["architecture"],
+                    type_=QuayClient.MANIFEST_V2S2_TYPE,
+                )
+            )
+        return digests
+
+    def _extract_manifest(self, repo, manifest, mtype):
+        hasher = hashlib.sha256()
+        hasher.update(manifest.encode("utf-8"))
+        digest = hasher.hexdigest()
+        return ManifestArchDigest(manifest=manifest, digest=digest, arch="amd64", type_=mtype)
+
+    _MEDIA_TYPES_PROCESS = {
+        QuayClient.MANIFEST_LIST_TYPE: _extract_ml_manifest,
+        QuayClient.MANIFEST_V2S2_TYPE: _extract_manifest,
+        QuayClient.MANIFEST_V2S1_TYPE: _extract_manifest,
+    }
+
+    def extract_manifests(self, image_ref, media_types, tolerate_missing=True):
+        """Extract manifests from container registry.
+
+        Method fetches manifests for all provided media types. If HTTPErrors is raised when
+        fetching for manifest and status code is 404 or 401, it's retried as there can be delay
+        in registry if manifest was just pushed to it.
+        When manifest is not found for given media type, it's not included in the result.
+
+        Args:
+            image_ref (str): Image reference in format <registry>/<repo>:<tag>
+            media_types (list): List of media types to be extracted.
+            tolerate_missing (bool): If True, missing manifests are tolerated and
+            empty list is returned.
+
+        Returns:
+            list: List of ManifestArchDigest objects.
+        """
+        mtypes = media_types or self._MEDIA_TYPES_PRIORITY.keys()
+        results = []
+        for mtype in mtypes:
+            for i in range(self.timeout // self.poll_rate):
+                try:
+                    manifest = self.quay_client.get_manifest(image_ref, media_type=mtype, raw=True)
+                except requests.exceptions.HTTPError as e:
+                    # When robot account is used, 401 may be returned instead of 404
+                    if (
+                        e.response.status_code == 404 or e.response.status_code == 401
+                    ) and tolerate_missing:
+                        manifest = None
+                    else:
+                        raise
+                except ManifestTypeError:
+                    manifest = None
+                    break
+                else:
+                    break
+                time.sleep(self.sleep_time)
+            if not manifest:
+                continue
+            else:
+                ret = self._MEDIA_TYPES_PROCESS[mtype](self, image_ref, manifest, mtype)
+                if isinstance(ret, list):
+                    results.extend(ret)
+                else:
+                    results.append(ret)
+        return results
```

Comment:
That would work if you want to fetch for all media types every time, but it's also used in the case when we fetch for new manifests which are not signed after push item is pushed to quay. So in that case it would complicate the method. So I think it's better to keep it more simple even when it can do extra call to the registry.
For tolerate_missing, there isn't any use-case for it yet, but I think it shouldn't be hardcoded, it would limit the usage of the method for the future